src/core/net/ip6.cpp

```diff
@@ -460,7 +460,15 @@ otError Ip6::SendDatagram(Message &aMessage, MessageInfo &aMessageInfo, IpProto
     header.SetDscp(PriorityToDscp(aMessage.GetPriority()));
     header.SetPayloadLength(payloadLength);
     header.SetNextHeader(aIpProto);
-    header.SetHopLimit(aMessageInfo.mHopLimit ? aMessageInfo.mHopLimit : static_cast<uint8_t>(kDefaultHopLimit));
+
+    if (aMessageInfo.GetHopLimit() != 0 || aMessageInfo.ShouldApplyZeroHopLimit())
+    {
+        header.SetHopLimit(aMessageInfo.GetHopLimit());
```

Comment:
If `ShouldApplyZeroHopLimit()` is `true`, should we force IPv6 Hop Limit to 0 regardless of `GetHopLimit()` value?

Reply:
@jwhui To serve the purpose you mentioned, `ShouldApplyZeroHopLimit()` should be updated to `ShouldApplyUserHopLimit()`, which would require the user to not only specify the mHopLimit but also set the flag (something like mApplyUserHopLimit)

However the code here is by design(as agreed with @abtink with least impact on user experience since hoplimit 0 is unlikely to be used generally) , with the `||` only handle the special  hoplimit value 0 regarding whether to apply 0 (if  `ShouldApplyZeroHopLimit() `is true), otherwise use the default `kDefaultHopLimit` as previously.

Reply:
With the current logic, if `ShouldApplyZeroHopLimit()` is `true`, the IPv6 Hop Limit is set to `GetHopLimit()`. My question is whether the IPv6 Hop Limit should be set to 0 whenever `ShouldApplyZeroHopLImit()` is `true`.

Reply:
Updated as your suggestion to avoid confusion. Thanks.